torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,187 @@
+  // Master runs the store_thread
+  if (_rank == 0) {
+    _store_thread = std::thread(store_thread_daemon);
+    std::tie(_store_port, std::ignore) = load_master_env();
+    _store_addr = "localhost";
+  } else {
+    std::tie(_store_addr, _store_port) = load_worker_env();
+  }
+  _socket = connect(_store_addr, _store_port);
+}
+
+Store::~Store() {
+  // the 0 process has to stop the daemon
+  if (_rank == 0) {
+    send_value<query_type>(_socket, query_type::finish);
+    _store_thread.join();
+  }
+}
+
+void Store::set(const std::string& key, const std::vector<char>& data) {
+  send_value<query_type>(_socket, query_type::set);
+  send_string(_socket, key, true);
+  send_vector<char>(_socket, data);
+}
+
+std::vector<char> Store::get(const std::string& key) {
+  wait({key});
+  send_value<query_type>(_socket, query_type::get);
+  send_string(_socket, key);
+  return recv_vector<char>(_socket);
+}
+
+void Store::wait(const std::vector<std::string>& keys) {
+  send_value<query_type>(_socket, query_type::wait);
+  size_type nkeys = keys.size();
+  send_bytes<size_type>(_socket, &nkeys, 1, (nkeys > 0));
+  for (int i = 0; i < nkeys; i++) {
+    send_string(_socket, keys[i], (i != (nkeys - 1)));
+  }
+  // after sending the query, wait for a 'stop_waiting' reponse
+  query_type qr;
+  recv_bytes<query_type>(_socket, &qr, 1);
+  if (qr != query_type::stop_waiting)
+    throw std::runtime_error("stop_waiting response expected");
+}
+
+namespace {
+
+static void store_thread_daemon() {
+  port_type port;
+  rank_type world_size;
+  int socket;
+  std::tie(port, world_size) = load_master_env();
+  std::vector<int> sockets(world_size);
+  std::tie(socket, std::ignore) = listen(port);
+  // accept WORLD_SIZE connections
+  for (int i = 0; i < world_size; i++) {
+    int p_socket;
+    std::tie(p_socket, std::ignore) = accept(socket);
+    sockets[i] = p_socket;
+  }
+
+  // listen for requests
+  struct pollfd fds[world_size];
+  for (int i = 0; i < world_size; i++) {
+    fds[i].fd = sockets[i];
+    fds[i].events = POLLIN;   // we only read
+  }
+
+  // receive the queries
+  bool finished = false;
+  while (!finished) {
+    for (int i = 0; i < world_size; i++) fds[i].revents = 0;
+    SYSCHECK(::poll(fds, world_size, -1));
+    for (int i = 0; i < world_size; i++) {
+      if (fds[i].revents == 0)
+        continue;
+      if (fds[i].revents ^ POLLIN)
+        throw std::system_error(ECONNABORTED, std::system_category());
+      finished = query(i, sockets);
+      if (finished)
+        break;
+    }
+  }
+}
+
+/* 
+ * query communicates with the worker. The format
+ * of the query is as follows:
+ * type of query | size of arg1 | arg1 | size of arg2 | arg2 | ...
+ * or, in the case of wait
+ * type of query | number of args | size of arg1 | arg1 | ...
+ */
+static bool query(rank_type rank, const std::vector<int>& sockets) {
+  static store_type store_;
+  static std::vector<std::vector<std::string>> awaited_keys(sockets.size());
```

Comment:
srsly make it a map of vectors. Right now you'll do a lot of string comparisons vs a single hash